SignalShareExtension/utils/ShareAppExtensionContext.h

```diff
@@ -6,12 +6,15 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+// This is _NOT_ a singleton and will be instantiated each time that the SAE is used.
 @interface ShareAppExtensionContext : NSObject <AppContext>
 
 - (instancetype)init NS_UNAVAILABLE;
 
 - (instancetype)initWithRootViewController:(UIViewController *)rootViewController;
 
+- (void)setupEnvironment;
```

Comment:
/me puts on my architect astronaut suit  




"environment setup" entails a lot of things...

Much of the logic which sets up our environment is duplicated between AppDelegate and ShareViewController, with some divergence (currently documented as comments in the ShareViewController where it diverges).

That all makes pretty good sense, and is likely accurate given that we just copy and pasted and hacked and slashed. But I'm concerned that this is not going to be a good long term approach as this logic continues to change, we will be required to make sure to make changes in both files, and keep comments referencing intentionally missing code up to date in order for it to continue to make as much sense as it makes now.

**tldr; I think app setup should be extracted to a place where the shared logic can truly be shared and the divergences are implemented via inheritance or protocol conformance. (e.g. an AppSetup class)**

As I said earlier, I think AppContext should serve only as an extension-compatible adapter-pattern around UIApplication, and we should try to resist it becoming a junk drawer for helper methods needing to know about lots of other classes.

But I'm less concerned with precisely where shared "environment setup" logic lives, and mostly just want it *to be shared*. e.g. if you like this approach of putting more logic into the AppContext, maybe we could use swift protocol extensions to share code, or we could have AppContext be a base class instead of a protocol. IMO that would be preferable to having this touchy code duplicated.

![image](https://user-images.githubusercontent.com/217057/33727266-8291385e-db45-11e7-8830-d255cedb6a22.png)

Reply:
I agree that we want to:

* DRY
* Avoid god objects / "junk drawers".

Note that this logic isn't quite the same between the two apps.  Our notifications manager and call message handler are quite different in the SAE - they fail when used.  

I'm trying to avoid dragging the main app's notifications manager and call message handler into `SignalMessaging`: one requires WebRTC etc. and the other uses non-SAE-safe API.  So if we used an `AppSetup` class, we have to inject factory blocks for these - that feels like over-engineering.

This relates to a similar question I've been thinking about: can we do better by replacing `Environment` and `TextSecureKitEnv` (and perhaps our "shared" singletons) with a global "singleton registry".  

* We could easily mock/modify singletons in our tests.
* We would be more explicit about dependencies and creation/kickoff order.
* It might make it possible to let us explicitly teardown singletons - which might be quite valuable in the SAE.  We'd need to specify teardown order/procedure.

Anyhow, I've taken your suggestion and created an `AppSetup` class but it feels over-engineered to me.  I'd like to see if we can come up with a better approach.